src/pages/experiments/[experimentId]/data-exploration/components/heatmap/HeatmapPlot.jsx

```diff
@@ -121,83 +122,148 @@ const HeatmapPlot = (props) => {
     }
   }, [width]);
 
-  const downsample = (groupBy) => {
+  // Defines sorting order according to which group the cell belongs to
+  const positionInGrouping = (cellSetNames, cell) => {
+    let position = 0;
+
+    for (const { key: cellSetName } of cellSetNames) {
+      if (cellsInSet(cellSetName).has(cell)) {
+        return position;
+      }
+
+      position += 1;
+    }
+
+    // If not in grouping set it in the end
+    return position + 1;
+  };
+
+  const cellsInSet = (cellSetName) => {
+    return properties[cellSetName].cellIds;
+  };
+
+  ///////////////////////////////////////////////////////////////////////////////////////////
+  // Returns a set of all cells relevant to the experiment (so the ones belonging to one or more of the enabled sets and not hidden).
+  const getAllRelevantCellIds = (arrayOfcellSetNamesByGroupByKey) => {
     // Find all hidden cells.
     const hiddenCellIds = union(Array.from(hidden), properties);
 
-    // Get how many cells we are sampling.
-    let total = 0;
+    let allCellsInSets = new Set();
 
-    // Create an object for storing the cells grouped by `groupBy`.
-    const groupedCells = {};
+    Object.values(arrayOfcellSetNamesByGroupByKey).forEach((arrayOfSetNames) => {
+      // Add each set of cells into 
+      arrayOfSetNames.forEach((setNameObj) => {
+        const cellSet = cellsInSet(setNameObj.key);
+        // Union of allCellsInSets and cellSet
+        allCellsInSets = new Set([...allCellsInSets, ...cellSet]);
+      });
+    });
 
-    // Find the `groupBy` root node.
-    const rootNode = hierarchy.filter((clusters) => clusters.key === groupBy);
+    // Only work with non-hidden cells.
+    const shownCells = new Set([...allCellsInSets].filter(x => !hiddenCellIds.has(x)));
 
-    if (!rootNode.length) {
-      return [];
+    return shownCells;
+  }
+
+  const getCellSetNamesDict = (rootNodes) => {
+    const cellSetNamesByGroupByKey = {};
+    rootNodes.forEach((currentRootNode) => cellSetNamesByGroupByKey[currentRootNode.key] = currentRootNode.children);
+
+    return cellSetNamesByGroupByKey
+  };
```

Comment:
there are lots of small functions in this file that are about massaging the cell sets data in various ways and forms in order to make their data structure efficient. This function is an example, but there are other functions in the other heatmap components too. These functions should be extracted into a separate file and only imported and called in here and other files - this is not functionality specific to the heatmap plot and it does not fit for these cell set data manipulation functions to be here.

Reply:
In the end I am using some functions for sets, but since they are not particularly about cellSets and more about generic sets I will create a different file.